### PR TITLE
Add a script to mirror plugins from wordpress.org

### PR DESCRIPTION
**Pull request by snim2**
Currently we maintain two plugin libraries: one on GitHub and one on GitLab. The GitLab library is mirrored from wordpress.org and the GitHub mirror reflects the repositories in GitLab.

Eventually, we want to standardise on one library, which will be in GitHub. To do this, we need to ensure that both plugin libraries mirror wordpress.org.

This PR adds a new Ruby script to mirror plugins from wordpress.org. Specifically it:

* Uses the same algorithm we had in Whippet Racetrack
* Adds explicit error handling
* Adds more structured log output

The mirroring algorithm is roughly:

* For each plugin repository:
    * Use the wordpress.org REST API to find the latest version of the plugin as a semver string.
    * Use the GitHub REST API to find the latest tag in the mirror repo.
    * Compare the versions, if they are identical move to the next repository.
    * Clone the plugin repository from GitHub.
    * Download the latest version of the plugin as a .zip file from wordpress.org.
    * Unzip the plugin files _over_ the repo clone.
    * Add and commit the new files.
    * Tag the repo with the new version number and a major-mode version.
    * Push the new commits and the new tags back to the GitHub repo.

This is also roughly identical to the manual process we use to update pro- and premium plugins.

The PR also adds a new workflow for the new mirror script, which differs from the main workflow in these ways:

* Add `DRY_RUN` and related variables for testing
* Run exactly 4 times a day

Before we deprecate the old script, this needs to be re-written, the testing env vars need removing and the schedule should be adjusted back to the original schedule from the `main.yml` configuration.

## Testing

Copy `.env.wordpress-example` to `.env` and fill in the token. Check that `DRY_RUN` is set to `true`.

Either use `act`, as per the README, or just run `./mirror-wordpress-plugins --dry-run`.

### CODE REVIEW

**Comment by snim2** [comment]
So I think what needs to change here is:

1. Copy the main script so we can have the old script running and the new script running in `--dry-run` mode
2. Add a second workflow to run the new script in `--dry-run` mode
3. Probably stick all that in a new PR

File: mirror-wordpress-plugins [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    :latest_github_version, :latest_wordpress_version
+
+  @@wordpress_api_url = "https://api.wordpress.org/plugins/info/1.0/"
+
+  def initialize(name, github_url, default_branch)
+    @slug = name
+    @github_url = github_url
+    @default_branch = default_branch
+    @full_path_to_clone = File.join(Dir.pwd, @slug)
+    @wordpress_api_url = "#{@@wordpress_api_url}#{@slug}.json"
```
Comment (RobjS) [inline mirror-wordpress-plugins:40]: I think we might want to do `@slug.downcase` here, e.g. compare the output of

```
curl -s https://api.wordpress.org/plugins/info/1.0/wp-recaptcha
```

and

```
curl -s https://api.wordpress.org/plugins/info/1.0/wp-reCAPTCHA
```

File: mirror-wordpress-plugins [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    puts "==> Failed to update the following #{@failed_updates.size} plugins:"
+    @failed_updates.each do |name, error_message|
+      puts
+      puts "*** #{name} failed to update with message:"
+      puts error_message
+    end
+  end
+
+  def plugin_library_repos
+    puts "==> Fetching all GitHub plugin repositiories in the library..."
```
Comment (RobjS) [inline mirror-wordpress-plugins:237]: Typo: `repositiories`

**Review by RobjS — CHANGES_REQUESTED** [review]
This looks good to me, and the dry-run completed successfully. The report at the end is really useful, highlights how many closed plugins we still have!

Just a few minor tweaks.

File: mirror-wordpress-plugins (hunk 1 above)
Comment (snim2) [reply mirror-wordpress-plugins]: I'd assumed here that our repo names were slugs!

**Comment by snim2** [comment]
@RobjS OK, one blocker to making this live will be plugins where there is a pro version and a non-pro version with the same slug. This is what I have locally for Agolia:

```shell
❯ git tree-log
* 1cb84fd (HEAD -> master, tag: v3) v3.1.2
* 6aec9c8 (tag: v4.1.1, tag: v4, origin/master, origin/HEAD) wp-algolia.4.1.1.zip
* 520a7e5 (tag: v4.1.0) wp-algolia.4.1.0.zip
* b95a3ad (tag: v4.0.0) wp-algolia.4.0.0.zip
* 7646d29 (tag: v3.1.2) Initial commit
```

**Comment by snim2** [comment]
> @RobjS OK, one blocker to making this live will be plugins where there is a pro version and a non-pro version with the same slug. This is what I have locally for Agolia:
> 
> ```shell
> ❯ git tree-log
> * 1cb84fd (HEAD -> master, tag: v3) v3.1.2
> * 6aec9c8 (tag: v4.1.1, tag: v4, origin/master, origin/HEAD) wp-algolia.4.1.1.zip
> * 520a7e5 (tag: v4.1.0) wp-algolia.4.1.0.zip
> * b95a3ad (tag: v4.0.0) wp-algolia.4.0.0.zip
> * 7646d29 (tag: v3.1.2) Initial commit
> ```



> @RobjS OK, one blocker to making this live will be plugins where there is a pro version and a non-pro version with the same slug. This is what I have locally for Agolia:
> 
> ```shell
> ❯ git tree-log
> * 1cb84fd (HEAD -> master, tag: v3) v3.1.2
> * 6aec9c8 (tag: v4.1.1, tag: v4, origin/master, origin/HEAD) wp-algolia.4.1.1.zip
> * 520a7e5 (tag: v4.1.0) wp-algolia.4.1.0.zip
> * b95a3ad (tag: v4.0.0) wp-algolia.4.0.0.zip
> * 7646d29 (tag: v3.1.2) Initial commit
> ```

I think probably the fix for this is to check that the new tag is "larger" than the old one, not just different.

**Review by RobjS — APPROVED** [review]
This is working for me, and all looks good. I don't think we can rely on pro plugins being ahead of their free equivalents, so approving on the basis that we'll find an alternative solution for that issue.

**Pull request merged**